ddtrace/contrib/anthropic/utils.py

```diff
@@ -0,0 +1,284 @@
+        content_block = chunk.delta
+        if getattr(content_block, "type", "") == "text_delta":
+            chunk_content = getattr(content_block, "text", "")
+            if chunk_content:
+                message["content"] += chunk_content
+
+        elif getattr(chunk, "type", "") == "message_delta":
+            # message delta events signal the end of the message
+            chunk_finish_reason = getattr(content_block, "stop_reason", "")
+            if chunk_finish_reason:
+                message["finish_reason"] = content_block.stop_reason
+                message["content"] = message["content"].strip()
+
+                chunk_usage = getattr(chunk, "usage", {})
+                if chunk_usage:
+                    message_usage = message.get("usage", {"output": 0, "input": 0})
+                    message_usage["output"] += getattr(chunk_usage, "output_tokens", 0)
+                    message_usage["input"] += getattr(chunk_usage, "input_tokens", 0)
+                    message["usage"] = message_usage
+                message_finished = True
+
+    return message, message_finished
+
+
+def _tag_streamed_chat_completion_response(integration, span, messages):
+    """Tagging logic for streamed chat completions."""
+    if messages is None:
+        return
+    for idx, message in enumerate(messages):
+        span.set_tag_str("anthropic.response.completions.%d.content" % idx, integration.trunc(message["content"]))
+        span.set_tag_str("anthropic.response.completions.%d.role" % idx, message["role"])
+        if message.get("finish_reason", None) is not None:
+            span.set_tag_str("anthropic.response.completions.%d.finish_reason" % idx, message["finish_reason"])
+        if message.get("usage", None) is not None:
+            span.set_metric("anthropic.response.completions.%d.usage.input" % idx, message["usage"].get("input", 0))
+            span.set_metric("anthropic.response.completions.%d.usage.output" % idx, message["usage"].get("output", 0))
+
+
+def _loop_handler(span, chunk, streamed_chunks):
+    """Sets the anthropic model tag and appends the chunk to the correct index in the streamed_chunks list.
+
+    When handling a streamed chat/completion response, this function is called for each chunk in the streamed response.
+    """
+    if span.get_tag("anthropic.response.model") is None:
+        span.set_tag("anthropic.response.model", chunk.message.model)
+    streamed_chunks.append(chunk)
+
+
+def _is_generator(resp):
+    # type: (...) -> bool
+    import anthropic
+
+    if isinstance(resp, Generator):
+        return True
+    if hasattr(anthropic, "Stream") and isinstance(resp, anthropic.Stream):
+        return True
+    return False
+
+
+def _is_async_generator(resp):
+    # type: (...) -> bool
+    import anthropic
+
+    if isinstance(resp, AsyncGenerator):
+        return True
+    if hasattr(anthropic, "AsyncStream") and isinstance(resp, anthropic.AsyncStream):
+        return True
+    return False
```

Comment:
Curious, does Anthropic use AsyncGenerator/Generator objects? From what I've seen it looks like they only use AsyncStream/Stream. This corresponding code from the openai integration was checking for AsyncGenerator/Generators to provide backwards support for older versions, which we don't need to do for anthropic. (unless it does return custom Generator/AsyncGenerator objects)

Reply:
I don't think so...? This is from the code I took from openai, so generator can be removed.

Reply:
Removed the generator checks